### PR TITLE
:rotating_light: [`Clip`] Fix masking and enable flash attention on all model types

### LINKED ISSUE #41668: CLIP incompatible with Flash Attention 3
### System Info

- `transformers` version: 4.57.0
- Platform: Linux-6.14.0-33-generic-x86_64-with-glibc2.39
- Python version: 3.12.3
- Huggingface_hub version: 1.0.0.rc4

### Who can help?

@ArthurZucker @Cyrilvallez

### Information

- [ ] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

```
from diffusers import EulerDiscreteScheduler, StableDiffusionXLPipeline
import torch
import os

model_path =os.getenv('MODEL_PATH')
print(model_path)
pipe = StableDiffusionXLPipeline.from_pretrained(
    model_path,
    variant="fp16",
    torch_dtype=torch.float16
).to("cuda")
pipe.fuse_qkv_projections()

pipe.text_encoder.config._attn_implementation = "flash_attention_3"
pipe.text_encoder_2.config._attn_implementation = "flash_attention_3"

prompt = ["A city at night with people walking around."] * 8
image = pipe(prompt, num_inference_steps=1).images[0]
```

Input into `_upad_input`
```
Starting with query_length: 77
query_layer shape: torch.Size([8, 77, 12, 64])
key_layer shape: torch.Size([8, 77, 12, 64])
value_layer shape: torch.Size([8, 77, 12, 64])
attention_mask shape: torch.Size([8, 1, 77, 77])
```

output from `_upad_input`

```
q shape: torch.Size([23408, 12, 64])
k shape: torch.Size([23408, 12, 64])
v shape: torch.Size([23408, 12, 64])
cu_seq_lens_q: torch.Size([8, 1, 78])
cu_seq_lens_k: torch.Size([8, 1, 78])
max_length_q: 0
max_length_k: 0

```

### Expected behavior

I'm trying to enable Flash Attention 3 for the CLIP encoder in SDXL. 

Is the line in the forward pass supposed to include FA3 as well?

https://github.com/huggingface/transformers/blob/19df66dcba7690fab0b82105c154685c63016a21/src/transformers/models/clip/modeling_clip.py#L322

Currently when the attention_implementation is set to FlashAttention3 the attention_mask is set to the the causal_attention_mask which is 4D and when passed into [_upad_input](https://github.com/huggingface/transformers/blob/19df66dcba7690fab0b82105c154685c63016a21/src/transformers/modeling_flash_attention_utils.py#L237) which only supports a 2D attention_mask leads to garbage output https://github.com/huggingface/transformers/blob/19df66dcba7690fab0b82105c154685c63016a21/src/transformers/modeling_flash_attention_utils.py#L258

### PR DESCRIPTION
**Pull request by vasqu**
Clip used old mask APIs leading to a confused usage:
- A causal mask (normal triu mask)
- A padding mask (encoder mask == only accounting for padding)
- Add both of above == final mask --> causal mask with padding

^ works only for interfaces with support for 4D masks which disabled FA usage in general.

This PR now correctly changes this to the new API which handles padding automatically. We have to additionally pass the `is_causal` kwarg to dynamically switch between modality types (text == causal, image == full). This is only enabled through recent PRs (fa #39707, sdpa #41692).

Closes #41673
Fixes #41668

### CODE REVIEW

**Comment by vasqu** [comment]
cc @yonigozlan when you come across models like these in the vision refactors

File: src/transformers/models/clip/modeling_clip.py [hunk 1]
```diff
@@ -604,23 +588,20 @@ def forward(
...
-            # [batch_size, seq_len] -> [batch_size, 1, tgt_seq_len, src_seq_len]
-            attention_mask = _prepare_4d_attention_mask(attention_mask, hidden_states.dtype)
-
         encoder_outputs: BaseModelOutput = self.encoder(
             inputs_embeds=hidden_states,
             attention_mask=attention_mask,
-            causal_attention_mask=causal_attention_mask,
             output_attentions=output_attentions,
             output_hidden_states=output_hidden_states,
+            is_causal=True,
```
Comment (vasqu) [inline src/transformers/models/clip/modeling_clip.py:556]: This is important to enable proper causal usage, dynamic switching. I masked with kwargs but could also properly type it out in the modules above.

File: src/transformers/models/clip/modeling_clip.py [hunk 2]
```diff
@@ -425,7 +415,7 @@ class CLIPPreTrainedModel(PreTrainedModel):
     _supports_sdpa = True
     _supports_flash_attn = True
     _supports_flex_attn = True
-    _supports_attention_backend = True
+    _supports_attention_backend = False  # kwargs are not supported throughout all modules
```
Comment (zucchini-nlp) [inline src/transformers/models/clip/modeling_clip.py]: maybe instead fix so it's propagated all along from model to attention layer?

File: src/transformers/models/clip/modeling_clip.py [hunk 3]
```diff
@@ -604,23 +588,20 @@ def forward(
...
-        # CLIP's text model uses causal mask, prepare it here.
-        # https://github.com/openai/CLIP/blob/cfcffb90e69f37bf2ff1e988237a0fbe41f33c04/clip/model.py#L324
-        causal_attention_mask = _create_4d_causal_attention_mask(
-            input_shape, hidden_states.dtype, device=hidden_states.device
+        attention_mask = create_causal_mask(
+            config=self.config,
+            input_embeds=hidden_states,
+            attention_mask=attention_mask,
+            cache_position=torch.arange(hidden_states.shape[1], device=hidden_states.device),
+            past_key_values=None,
```
Comment (zucchini-nlp) [inline src/transformers/models/clip/modeling_clip.py:549]: not clear why we create only causal mask tbh. Is it going to work with bidirectional mask for vision model? Let's also run slow tests before merging

File: src/transformers/models/clip/modeling_clip.py (hunk 1 above)
Comment (zucchini-nlp) [reply src/transformers/models/clip/modeling_clip.py:556]: Same q here, are we not supposed to set `is_causal=False` when the attention is used within a `VisionModel`?

File: src/transformers/models/clip/modeling_clip.py (hunk 2 above)
Comment (vasqu) [reply src/transformers/models/clip/modeling_clip.py]: Fair enough, I can properly use kwargs :D

File: src/transformers/models/clip/modeling_clip.py (hunk 3 above)
Comment (vasqu) [reply src/transformers/models/clip/modeling_clip.py:549]: In the text module of CLIP, it was always causal (but obfuscated by mixing old and new ish APIs). 
In the vision modules of CLIP, no mask is used at all (full attention).

For sure gonna check slow tests

File: src/transformers/models/clip/modeling_clip.py (hunk 1 above)
Comment (vasqu) [reply src/transformers/models/clip/modeling_clip.py:556]: It will use the modules's attribute (which is False) in the vision parts. For the text parts, we overwrite with the kwarg instead.

File: src/transformers/models/clip/modeling_clip.py (hunk 3 above)
Comment (zucchini-nlp) [reply src/transformers/models/clip/modeling_clip.py:549]: oh my bad, this is a `TextTransformer`, makes sense

**Review by zucchini-nlp — APPROVED** [review]
Didn't see that we're changing only the text model. LGTM, as long as the slow tests are passing

File: src/transformers/models/clip/modeling_clip.py [hunk 4]
```diff
@@ -303,8 +303,8 @@ def forward(
         self,
         hidden_states: torch.Tensor,
         attention_mask: Optional[torch.Tensor] = None,
-        causal_attention_mask: Optional[torch.Tensor] = None,
         output_attentions: Optional[bool] = False,
+        **kwargs,
```
Comment (molbap) [inline src/transformers/models/clip/modeling_clip.py]: so here shouldn't the kwargs be typed? It's my only comment :D

File: src/transformers/models/clip/modeling_clip.py [hunk 5]
```diff
@@ -422,10 +406,15 @@ class CLIPPreTrainedModel(PreTrainedModel):
     base_model_prefix = "clip"
     input_modalities = ["image", "text"]
     supports_gradient_checkpointing = True
+    accepts_loss_kwargs = False
```
Comment (vasqu) [inline src/transformers/models/clip/modeling_clip.py]: Pretty sure that we do not use the proper loss fcts for this yet

File: src/transformers/models/clip/modeling_clip.py [hunk 6]
```diff
@@ -520,82 +506,43 @@ def forward(
...
-
         return BaseModelOutput(
             last_hidden_state=hidden_states,
-            hidden_states=encoder_states,
-            attentions=all_attentions,
         )
 
 
-class CLIPTextTransformer(nn.Module):
+class CLIPTextTransformer(CLIPPreTrainedModel):
```
Comment (vasqu) [inline src/transformers/models/clip/modeling_clip.py]: Due to some models using this as "model" backbone, I needed to make this a pretrained module to work with check_inputs

File: src/transformers/models/clip/modeling_clip.py [hunk 7]
```diff
@@ -520,82 +506,43 @@ def forward(
...
+        self.gradient_checkpointing = False
+
         self.embeddings = CLIPTextEmbeddings(config)
         self.encoder = CLIPEncoder(config)
         self.final_layer_norm = nn.LayerNorm(embed_dim, eps=config.layer_norm_eps)
 
         # For `pooled_output` computation
         self.eos_token_id = config.eos_token_id
 
+    @check_model_inputs(tie_last_hidden_states=False)
```
Comment (vasqu) [inline src/transformers/models/clip/modeling_clip.py]: Unsure, but seen this in other vision models

**Comment by vasqu** [comment]
run-slow: clip

**Comment by vasqu** [comment]
@molbap @zucchini-nlp I changed a few things to align the kwargs with our modern practices, i.e. see 764e63fa819d02d4e570e2529a36fd0050121114

This makes kwarg easy to type properly, otherwise we probably need to type it as FA kwargs :thinking:

File: src/transformers/models/clip/modeling_clip.py (hunk 6 above)
Comment (molbap) [reply src/transformers/models/clip/modeling_clip.py]: sounds good to me

File: src/transformers/models/clip/modeling_clip.py (hunk 7 above)
Comment (molbap) [reply src/transformers/models/clip/modeling_clip.py]: yes that should do it!

**Review by molbap — APPROVED** [review]
Much better with the typing, thanks!

File: src/transformers/models/clip/modeling_clip.py (hunk 7 above)
Comment (zucchini-nlp) [reply src/transformers/models/clip/modeling_clip.py]: it can be tested by running a VLM with CLIP-backbone. Basically in most vision models the `out.last_hidden_state != out.hidden_states[-1]`, therefore the flag is set to `False`

I think it's just some legacy behavior without strong reason, but since these models are used in VLMs we can't chnage it anymore without breaking generation

Comment (vasqu) [reply src/transformers/models/clip/modeling_clip.py]: Do you have a quick example model to check? :D But yea sounds reasonable, it is what it is. Let's keep the proper behavior

**Comment by vasqu** [comment]
CI has issues today, will probably check tomorrow again (and propogate the changes to metaclip 2 + mlcd) and merge then

File: src/transformers/models/clip/modeling_clip.py (hunk 7 above)
Comment (vasqu) [reply src/transformers/models/clip/modeling_clip.py]: Or rather should I check it? Is it safe the current way I did it

Comment (zucchini-nlp) [reply src/transformers/models/clip/modeling_clip.py]: I think simple Llava slow tests are enough

**Comment by vasqu** [comment]
run-slow: clip, metaclip_2, mlcd, llava

**Comment by vasqu** [comment]
run-slow: clip, metaclip_2, mlcd, llava

**Comment by vasqu** [comment]
Yea, the CI is not having a good day :D locally all the relevant tests passed, especially the integration tests - checking tomorrow

File: src/transformers/models/clip/modeling_clip.py (hunk 7 above)
Comment (vasqu) [reply src/transformers/models/clip/modeling_clip.py]: Checked locally and the tests passed. Will check tomorrow again, the CI is too wonky

**Comment by vasqu** [comment]
run-slow: clip, metaclip_2, mlcd, llava

**Comment by vasqu** [comment]
run-slow: clip, metaclip_2, mlcd, llava

File: tests/models/mlcd/test_modeling_mlcd.py [hunk 8]
```diff
@@ -146,7 +146,7 @@ class MLCDVisionModelIntegrationTest(unittest.TestCase):
     @slow
     def test_inference(self):
         model_name = "DeepGlint-AI/mlcd-vit-bigG-patch14-448"
-        model = MLCDVisionModel.from_pretrained(model_name).to(torch_device)
+        model = MLCDVisionModel.from_pretrained(model_name, attn_implementation="eager").to(torch_device)
```
Comment (vasqu) [inline tests/models/mlcd/test_modeling_mlcd.py:149]: Needed for output attns

**Comment by vasqu** [comment]
run-slow: clip, metaclip_2, mlcd, llava

File: src/transformers/models/clip/modeling_clip.py (hunk 7 above)
Comment (vasqu) [reply src/transformers/models/clip/modeling_clip.py]: Already handled by #41546 now. This PR really does only change the relevant attn parts atp

File: tests/models/mlcd/test_modeling_mlcd.py (hunk 8 above)
Comment (molbap) [reply tests/models/mlcd/test_modeling_mlcd.py:149]: makes sense

File: src/transformers/models/clip/modeling_clip.py [hunk 9]
```diff
@@ -344,13 +334,12 @@ def forward(
...
             values,
             attention_mask,
-            is_causal=self.is_causal,
             scaling=self.scale,
             dropout=0.0 if not self.training else self.dropout,
             **kwargs,
         )
 
-        attn_output = attn_output.reshape(batch_size, seq_length, embed_dim).contiguous()
+        attn_output = attn_output.reshape(batch_size, seq_length, -1).contiguous()
```
Comment (ArthurZucker) [inline src/transformers/models/clip/modeling_clip.py:342]: usually we use -1 for the batchsize as text can be ragged but not an issue

File: src/transformers/models/clip/modeling_clip.py [hunk 10]
```diff
@@ -632,8 +611,7 @@ def get_input_embeddings(self) -> nn.Module:
     def set_input_embeddings(self, value):
         self.text_model.embeddings.token_embedding = value
 
-    @check_model_inputs()
-    @can_return_tuple
```
Comment (ArthurZucker) [inline src/transformers/models/clip/modeling_clip.py:636]: indeed

**Review by ArthurZucker — APPROVED** [review]
thanlks
